### PR TITLE
STP refactor and unified stream

### PR DESCRIPTION
**Pull request by nicklester**

### CODE REVIEW

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/sparql/SparqlTriggeredProcessor.scala [hunk 1]
```diff
@@ -127,27 +102,103 @@ class SparqlTriggeredProcessor(config: Config,
...
+      Future
+        .sequence(processedSensors)
+        .map(updatedSensors => c.copy(sensors = updatedSensors))
+    }
+  }
+
+
+
+
+  def createAgentSource(config: Config,
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/sparql/SparqlTriggeredProcessor.scala]: Unused empty def removed

File: server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/SparqlProcessorManager.scala [hunk 2]
```diff
@@ -108,21 +111,58 @@ class SparqlProcessorManager(settings: SparqlProcessorManagerSettings) extends A
...
+    .groupedWeightedWithin((25*1024), 10.seconds)(_._1.size)
+    .via(
+      Ingester.ingesterFlow(baseUrl = settings.hostWriteOutput,
+        format = settings.materializedViewFormat,
+        writeToken = Option(settings.writeToken),
+        extractContext = (context: Option[StpMetadata]) => context match {
+          case Some(stpMetadata) => IngesterRuntimeConfig(stpMetadata.agentConfig.force.getOrElse(false),
+            stpMetadata.agentConfig.name)
+        },
+        connectionPool = Some(Retry.createNewHostConnectionPool[StpMetadata](settings.hostWriteOutput))
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/SparqlProcessorManager.scala]: @e-orz Once this is materialised there will only be one creation of the connection pool for Ingestion I think - do you not think that is the case?

File: server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/SparqlProcessorManager.scala [hunk 3]
```diff
@@ -108,21 +111,58 @@ class SparqlProcessorManager(settings: SparqlProcessorManagerSettings) extends A
...
   type Jobs = Map[String, JobStatus]
 
+  var currentJobs: Jobs = Map.empty
+  var configMonitor: Cancellable = _
+
   //todo: should we use injected ec??? implicit val ec = context.dispatcher
   implicit val system: ActorSystem = context.system
   implicit val mat = ActorMaterializer()
 
+  type ConnectionPool = Flow[(HttpRequest,State[(Option[StpMetadata],Long)]),
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/SparqlProcessorManager.scala]: Good idea, done

File: server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/SparqlProcessorManager.scala [hunk 4]
```diff
@@ -108,21 +111,58 @@ class SparqlProcessorManager(settings: SparqlProcessorManagerSettings) extends A
...
+    (Try[HttpResponse], State[(Option[StpMetadata],Long)]), Http.HostConnectionPool]
+
+  var connectionPools : Map[String,ConnectionPool] = Map.empty
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-sparql-agent/src/main/scala/cmwell/tools/data/sparql/SparqlProcessorManager.scala]: Good idea, done

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/downloader/consumer/BufferedTsvSource.scala [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+  def isHorizon(consumeComplete: Boolean, buf: mutable.Queue[Option[(Token, TsvData)]]) = consumeComplete && buf.isEmpty
+
+  def decodeToken(token: Token) = Try(
+    new org.joda.time.LocalDateTime(
+      Tokens.decompress(token).takeWhile(_ != '|').toLong
+    )
+  )
+
+  private val HttpAddress(protocol, host, port, _) = ArgsManipulations.extractBaseUrl(baseUrl)
+  private val conn = HttpConnections.newHostConnectionPool[Option[_]](host, port, protocol)
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/downloader/consumer/BufferedTsvSource.scala]: @e-orz Not at present. As there is only one consumer per host and there is not a particular high throughput on this part of the flow I have left it independent at present. This doesn't affect the overall throttling/backpressuring of the stream

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala [hunk 6]
```diff
@@ -31,14 +33,26 @@ import cmwell.util.akka.http.HttpZipDecoder
...
-import scala.util.{Failure, Success, Try}
 import scala.language.implicitConversions
-
 import scala.reflect.runtime.universe._
+import scala.util.{Failure, Success, Try}
 
 
 object Retry extends DataToolsLogging with DataToolsConfig {
 
+  def createNewHostConnectionPool[T](hostName: String)
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala]: @e-orz Yes - by SparqlProcessorManager

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/ingester/Ingester.scala [hunk 7]
```diff
@@ -17,38 +17,130 @@ package cmwell.tools.data.ingester
...
+      val req = HttpRequest(uri = uri, method = HttpMethods.POST, entity = entity,
+        headers = scala.collection.immutable.Seq(gzipContentEncoding))
+
+      writeToken match {
+        case Some(token) => req.addHeader(RawHeader("X-CM-WELL-TOKEN", token))
+        case None        => req
+      }
+    }
+
+    val parallelism = config.getInt("akka.http.host-connection-pool.max-connections")
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/ingester/Ingester.scala]: Good idea resolved.

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/ingester/Ingester.scala [hunk 8]
```diff
@@ -17,38 +17,130 @@ package cmwell.tools.data.ingester
...
+      .map {
+        case (Failure(ex), infotons, stpMetadata) =>
+          logger.error(s"${getLabel(stpMetadata)} problem: ${ex}")
+          badDataLogger.info(s"${getLabel(stpMetadata)} data: ${infotons.map(_.utf8String).mkString("\n")}")
+          (IngestFailEvent(numInfotons = infotons.size), stpMetadata)
+
+        case (Success(res @ HttpResponse(s, h, e, p)), infotons, stpMetadata) if s.isFailure() =>
+          logger.error(s"${getLabel(stpMetadata)} problem: host=${getHostname(h)} $e, $p")
+          badDataLogger.info(s"${getLabel(stpMetadata)} data: ${infotons.map(_.utf8String).mkString("\n")}")
+          res.discardEntityBytes()
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/ingester/Ingester.scala]: Good idea resolved

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/sparql/SparqlProcessor.scala [hunk 9]
```diff
@@ -181,114 +254,102 @@ class SparqlProcessor[T](baseUrl: String,
...
+        contextAndStartTime -> Source.empty
+      case (Failure(err), paths, contextAndStartTime) =>
+        redLogger.error(s"error: $err paths=${paths.map(_.utf8String).mkString(",")}", err)
+        contextAndStartTime -> Source.empty
+
+      case x =>
+        logger.error(s"unexpected message: $x")
+        None -> Source.empty
+    }
+    .mapAsyncUnordered(httpParallelism) {
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/sparql/SparqlProcessor.scala]: @e-orz This is original code if I remember correctly, I don't think I have changed this

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/sparql/SparqlProcessor.scala [hunk 10]
```diff
@@ -181,114 +254,102 @@ class SparqlProcessor[T](baseUrl: String,
...
+      case (Some((context, startTime)), dataLines) =>
+        logger.debug("received response from sparql and startTime={}", startTime)
+        Future.successful(dataLines -> context)
+    }
+    .mapAsyncUnordered(httpParallelism) {
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/sparql/SparqlProcessor.scala]: @e-orz This is original code if I remember correctly, I don't think I have changed this

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/downloader/consumer/BufferedTsvSource.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/downloader/consumer/BufferedTsvSource.scala]: @e-orz Reccommend we can address this in a later PR

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/ingester/Ingester.scala [hunk 11]
```diff
@@ -128,9 +128,12 @@ object Ingester extends DataToolsLogging with DataToolsConfig {
           (IngestFailEvent(numInfotons = infotons.size), stpMetadata)
 
         case (Success(res @ HttpResponse(s, h, e, p)), infotons, stpMetadata) if s.isFailure() =>
-          logger.error(s"${getLabel(stpMetadata)} problem: host=${getHostname(h)} $e, $p")
-          badDataLogger.info(s"${getLabel(stpMetadata)} data: ${infotons.map(_.utf8String).mkString("\n")}")
-          res.discardEntityBytes()
+          e.toStrict(1.minute).map { strict =>
+            logger.error(s"${getLabel(stpMetadata)} problem: host=${getHostname(h)} ${strict.data}, $p")
+            badDataLogger.info(s"${getLabel(stpMetadata)} data: ${infotons.map(_.utf8String).mkString("\n")}")
+            res.discardEntityBytes()
```
Comment (e-orz) [inline server/cmwell-data-tools/src/main/scala/cmwell/tools/data/ingester/Ingester.scala]: There is no reason to discard the entity bytes if you already converted it to strict.

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/sparql/SparqlProcessor.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/sparql/SparqlProcessor.scala]: ok.

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/sparql/SparqlProcessor.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (e-orz) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/sparql/SparqlProcessor.scala]: ok.

File: server/cmwell-data-tools/src/test/scala/cmwell/tools/data/sparql/SparqlTriggeredProcessorSpec.scala [hunk 12]
```diff
@@ -19,22 +19,18 @@ package cmwell.tools.data.sparql
...
 import com.github.tomakehurst.wiremock.stubbing.Scenario
-import org.scalatest.{Ignore, Matchers}
+import org.scalatest.Ignore
 
-import scala.concurrent.Future
+import scala.collection.JavaConverters._
 import scala.concurrent.duration._
-import collection.JavaConverters._
-
+/*
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-data-tools/src/test/scala/cmwell/tools/data/sparql/SparqlTriggeredProcessorSpec.scala]: as discussed the long disabled tests will follow in a subsequent PR

**Pull request merged**